evm/src/fuzz/strategies/state.rs

```diff
@@ -0,0 +1,99 @@
+use super::fuzz_param_from_state;
+use crate::executor::StateChangeset;
+use bytes::Bytes;
+use ethers::{
+    abi::Function,
+    types::{H256, U256},
+};
+use proptest::prelude::{BoxedStrategy, Strategy};
+use revm::{opcode, spec_opcode_gas, SpecId};
+use std::{cell::RefCell, collections::HashSet, io::Write, rc::Rc};
+
+/// A set of arbitrary 32 byte data from the VM used to generate values for the strategy.
+///
+/// Wrapped in a shareable container.
+pub type EvmFuzzState = Rc<RefCell<HashSet<[u8; 32]>>>;
+
+/// Given a function and some state, it returns a strategy which generated valid calldata for the
+/// given function's input types, based on state taken from the EVM.
+pub fn fuzz_calldata_from_state(
+    func: Function,
+    state: EvmFuzzState,
+) -> BoxedStrategy<ethers::types::Bytes> {
+    let strats = func
+        .inputs
+        .iter()
+        .map(|input| fuzz_param_from_state(&input.kind, state.clone()))
+        .collect::<Vec<_>>();
+
+    strats
+        .prop_map(move |tokens| {
+            tracing::trace!(input = ?tokens);
+            func.encode_input(&tokens).unwrap().into()
+        })
+        .no_shrink()
+        .boxed()
+}
+
+/// Collects state changes from a [StateChangeset] into an [EvmFuzzState].
+pub fn collect_state_from_changeset(state_changeset: &StateChangeset, state: EvmFuzzState) {
+    let state = &mut *state.borrow_mut();
+
+    // TODO: Old values
+    // TODO: Logs
+    for (address, account) in state_changeset {
+        // Insert basic account information
+        state.insert(H256::from(*address).into());
+        state.insert(u256_to_h256(account.info.balance).into());
+        state.insert(u256_to_h256(U256::from(account.info.nonce)).into());
+
+        // Insert storage
+        for (slot, value) in &account.storage {
+            state.insert(u256_to_h256(*slot).into());
+            state.insert(u256_to_h256(*value).into());
+        }
+
+        // Insert push bytes
+        if let Some(code) = &account.info.code {
+            for push_byte in collect_push_bytes(code.clone()) {
+                state.insert(push_byte);
+            }
+        }
+    }
+}
+
+/// Collects all push bytes from the given bytecode.
+fn collect_push_bytes(code: Bytes) -> Vec<[u8; 32]> {
```

Comment:
Currently panics on some solmate tests. My current thoughts on why is because some tests deploy random `bytes` as contracts and there is no guarantee that a `PUSH` opcode is actually well-formed

Reply:
That's fixed now, but this slows down solmate **a lot** because some of the tests deploy very large blobs of bytes. Should maybe impose a size limit here?

Reply:
Yeah @brockelmore wdyt, did you try benchmarking against Solmate?

Reply:
@onbjerg on further thought I think limiting the push bytes we iterate over the 24KB might be the right way. Another thing that would be useful is if we could collect the push bytes without a loop?

Reply:
Only way I can think of getting push bytes outside of a loop is running with debugging on

Reply:
Maybe we could do a regex? Would that be faster? @mattsse